Frameworks/CoreGraphics/CGContext.mm

```diff
@@ -574,88 +660,107 @@ void CGContextRestoreGState(CGContextRef context) {
-
     ComPtr<ID2D1CommandList> commandList;
     FAIL_FAST_IF_FAILED(deviceContext->CreateCommandList(&commandList));
+
+    deviceContext->BeginDraw();
     deviceContext->SetTarget(commandList.Get());
 
-    CGAffineTransform& currentTransform = currentState.transform;
-    CGAffineTransform transform = CGAffineTransformScale(currentTransform, 1.0, -1.0);
-    transform = CGAffineTransformTranslate(transform, 0., targetSize.height);
-    deviceContext->SetTransform({ transform.a, transform.b, transform.c, transform.d, transform.tx, transform.ty });
+    std::forward<Lambda>(drawLambda)(context, deviceContext.Get());
 
-    if (drawMode & kCGPathFill) {
-        if (drawMode & kCGPathEOFill) {
-            // TODO(DH): GH#1077 Regenerate geometry in Even/Odd fill mode.
-        }
-        deviceContext->FillGeometry(geometry, currentState.fillBrush.Get());
-    }
+    FAIL_FAST_IF_FAILED(deviceContext->EndDraw());
+    FAIL_FAST_IF_FAILED(commandList->Close());
 
-    if (drawMode & kCGPathStroke) {
-        ComPtr<ID2D1Factory> factory;
-        deviceContext->GetFactory(&factory);
-
-        // TODO(DH): GH#1077 Do not recreate for every drawing operation
-        ComPtr<ID2D1StrokeStyle> strokeStyle;
-        // dashes must be adjusted to be based on line width
-        std::vector<CGFloat> adjustedDashes{ currentState.dashes };
-        for (float& f : adjustedDashes) {
-            f /= currentState.lineWidth;
-        }
-        FAIL_FAST_IF_FAILED(
-            factory->CreateStrokeStyle(currentState.strokeProperties, adjustedDashes.data(), adjustedDashes.size(), &strokeStyle));
+    deviceContext->SetTarget(originalTarget.Get());
 
-        deviceContext->DrawGeometry(geometry, currentState.strokeBrush.Get(), currentState.lineWidth, strokeStyle.Get());
-    }
+    *outCommandList = commandList.Detach();
+}
 
-    deviceContext->EndDraw();
-    commandList->Close();
+static void __CGContextRenderCommandList(CGContextRef context, ID2D1CommandList* commandList) {
+    ComPtr<ID2D1RenderTarget> renderTarget = context->RenderTarget();
+    auto& state = context->CurrentGState();
+
+    ComPtr<ID2D1DeviceContext> deviceContext;
+    FAIL_FAST_IF_FAILED(renderTarget.As(&deviceContext));
+
+    D2D1_SIZE_F targetSize = renderTarget->GetSize();
+    CGAffineTransform& currentTransform = state.transform;
+
+    // CG is anchored in the bottom left, but D2D is anchored in the top left.
+    // Flip the context.
+    CGAffineTransform transform = CGAffineTransformScale(currentTransform, 1.0, -1.0);
+    // Translate it back onscreen.
+    transform = CGAffineTransformTranslate(transform, 0., targetSize.height);
+    deviceContext->SetTransform(__CGAffineTransformToD2D_F(transform));
 
     deviceContext->BeginDraw();
-    deviceContext->SetTarget(originalTarget.Get());
 
     bool layer = false;
-    if (currentState.alpha != 1.0f || false /* mask, clip, etc. */) {
+    if (state.alpha != 1.0f || false /* mask, clip, etc. */) {
```

Comment:
This is covered by upcoming tasks.

---
In reply to: [84129211](https://github.com/Microsoft/WinObjC/pull/1181#discussion_r84129211) [](ancestors = 84129211)